services/src/test/java/edu/unc/lib/dl/cdr/services/processing/AddContainerServiceTest.java

```diff
@@ -142,31 +158,65 @@ public void addCollectionToFolderTest() {
         }
     }
 
+    @SuppressWarnings("unchecked")
     @Test
     public void addFolderToCollectionTest() {
         CollectionObject collection = mock(CollectionObject.class);
         FolderObject folder = mock(FolderObject.class);
         when(repoObjLoader.getRepositoryObject(eq(parentPid))).thenReturn(collection);
         when(repoObjFactory.createFolderObject(any(ModelCom.class))).thenReturn(folder);
+        when(folder.getPid()).thenReturn(childPid);
         when(folder.getPremisLog()).thenReturn(premisLogger);
+        when(agent.getUsername()).thenReturn("some_valid_user");
+
+        Collection<PID> destinations = new ArrayList<>();
+        destinations.add(parentPid);
+        Collection<PID> added = new ArrayList<>();
+        added.add(childPid);
 
         service.addContainer(agent, parentPid, Cdr.Folder);
 
         verify(premisLogger).buildEvent(eq(Premis.Creation));
         verify(eventBuilder).write();
+
+        verify(messageSender).sendAddOperation(anyString(), destinationsCaptor.capture(),
```

Comment:
I'm trying to work through my feelings on what is important to have in unit tests or not, so don't take this as a recommendation.  After a number of times now in the fcrepo4 branch of having to go back and fix unit tests when implementation details change, I've been wanting to think through our testing strategies some.  Verifying that a message is sent does seem important to me, but it also makes the test significantly longer and more complicated.

I understand that in a unit test it is helpful to check all these things to catch programming errors.  On the other hand, are aspects like sending messages and recording events side effects/implementation details or actual expected outcomes of the method according to the public api?  It could be argued that they are since this is a preservation repository, but the explicit purpose of the class is to create containers.  Should we be testing that extra stuff like this happens in a unit test or an integration test?

I was reading a few articles about testing practices:
https://gist.github.com/Integralist/7944948
http://codebetter.com/iancooper/2011/10/06/avoid-testing-implementation-details-test-behaviours/
https://gist.github.com/androidfred/8210e67f37f5ca21da51e2fa5368853b (this one feels a bit contrived to me)

Anyways, this is probably fine (other than maybe using the `Arrays.asList` thing), it was just making me question my junit existence some more.